habitat/core/dataset.py

```diff
@@ -101,3 +103,124 @@ def default(self, object):
+
+    def filter_episodes(
+        self, filter_fn: Callable[[Episode], bool]
+    ) -> "Dataset":
+        """
+        Returns a new dataset with only the filtered episodes from the original
+        dataset.
+        Args:
+            filter_fn: Function used to filter the episodes.
+        Returns:
+            The new dataset.
+        """
+        new_episodes = []
+        for episode in self.episodes:
+            if filter_fn(episode):
+                new_episodes.append(episode)
+        new_dataset = copy.copy(self)
+        new_dataset.episodes = new_episodes
+        return new_dataset
+
+    def get_splits(
+        self,
+        num_splits: int,
+        max_episodes_per_split: Optional[int] = None,
+        remove_unused_episodes: bool = False,
+    ) -> List["Dataset"]:
+        """
+        Returns a list of new datasets, each with a subset of the original
+        episodes. All splits will have the same number of episodes.
+        Args:
+            num_splits: The number of splits to create.
+            max_episodes_per_split: If provided, each split will have up to this
+                many episodes. If it is not provided, each dataset will have
+                len(original_dataset.episodes) // num_splits episodes. If
+                max_episodes_per_split is provided and is larger than this
+                value, it will be capped to this value.
+            remove_unused_episodes: Once the splits are created, the extra
+                episodes will be destroyed from the original dataset. This saves
+                 memory for large datasets.
+        Returns:
+            A list of new datasets, each with their own subset of episodes.
+        """
+
+        assert (
+            len(self.episodes) >= num_splits
+        ), "Not enough episodes to create this many splits."
+
+        new_datasets = []
+        if max_episodes_per_split is None:
+            max_episodes_per_split = len(self.episodes) // num_splits
+        max_episodes_per_split = min(
+            max_episodes_per_split, (len(self.episodes) // num_splits)
+        )
+        if num_splits * max_episodes_per_split == len(self.episodes):
+            # Special case for exact size match
+            stride = len(self.episodes) // num_splits
+            for ii, split in enumerate(
+                range(0, len(self.episodes), stride)[:num_splits]
+            ):
+                new_dataset = copy.copy(self)  # Creates a shallow copy
+                new_dataset.episodes = new_dataset.episodes[
+                    split : split + stride
+                ].copy()
+                new_datasets.append(new_dataset)
+            new_episodes = self.episodes
+        else:
+            rand_items = np.random.choice(
+                len(self.episodes),
+                num_splits * max_episodes_per_split,
+                replace=False,
+            )
+            # Sort it so episode scene doesn't switch as often.
+            rand_items.sort()
```

Comment:
Maybe this responsibility should be delegated to the user for the dataset rather than us sorting and returning the data.

Reply:
I'll add a parameter for collating from the same scene_id.